File: lclang/src/main/java/lclang/statements/IfStatement.java
package lclang.statements;

import lclang.Caller;
import lclang.LCBaseExecutor;
import lclang.Link;
import lclang.exceptions.LCLangRuntimeException;
import lclang.libs.lang.classes.BoolClass;
import lclang.libs.lang.classes.LCClass;
import lclang.libs.lang.classes.VoidClass;
import lclang.statements.expressions.Expression;

public class IfStatement extends Statement {
    public final Expression condition;
    public final Statement ifTrue;
    public final Statement ifFalse;

    public IfStatement(Expression condition, Statement ifTrue, Statement ifFalse) {
        super(0);
        this.condition = condition;
        this.ifTrue = ifTrue;
        this.ifFalse = ifFalse;
    }

    @Override
    public Link visit(Caller prevCaller, LCBaseExecutor visitor) throws LCLangRuntimeException {
        LCClass cond = condition.visit(prevCaller, visitor).get.invoke(prevCaller);
        if(cond!=BoolClass.FALSE) {
            return ifTrue.visit(prevCaller, visitor);
        }

        if(ifFalse!=null){
            return ifFalse.visit(prevCaller, visitor);
        }

        return VoidClass.value;
    }
}

File: lclang/src/main/kotlin/lclang/statements/IfStatement.kt
package lclang.statements

import lclang.Caller
import lclang.LCBaseExecutor
import lclang.VOID_VALUE
import lclang.Value
import lclang.libs.lang.BoolClass

class IfStatement(
    val condition: Expression,
    val onTrue: Statement,
    val onFalse: Statement?,
    line: Int,
    column: Int
): Statement(line, column) {
    override fun visit(prevCaller: Caller, visitor: LCBaseExecutor): Value {
        val caller = getCaller(prevCaller)
        val cond = condition.visit(caller, visitor).get(caller)
        if(cond!= BoolClass.FALSE) {
            return onTrue.visit(caller, visitor)
        }

        return onFalse?.visit(caller, visitor) ?: VOID_VALUE
    }
}